### PR TITLE
GH-115651: Convert `LOAD_MODULE_ATTR` into `LOAD_INLINE_CONST` when the module is itself a constant.

### PR DESCRIPTION
**Pull request by markshannon**
Module attributes are global variables from other modules. It makes sense to convert them to constants as we already do for global variables.

* Issue: gh-115651

### CODE REVIEW

File: Python/optimizer_analysis.c [hunk 1]
```diff
@@ -670,7 +686,7 @@ remove_unneeded_uops(_PyUOpInstruction *buffer, int buffer_size)
      * could error. _CHECK_VALIDITY is needed if the previous
      * instruction could have escaped. */
     int last_set_ip = -1;
-    bool may_have_escaped = false;
+    bool may_have_escaped = true;
```
Comment (markshannon) [inline Python/optimizer_analysis.c:680]: This is not strictly related to this PR, but is generally necessary.

File: Python/optimizer_analysis.c [hunk 2]
```diff
@@ -276,6 +276,20 @@ sym_is_null(_Py_UOpsSymType *sym)
     return (sym->flags & (IS_NULL | NOT_NULL)) == IS_NULL;
 }
 
+static inline bool
```
Comment (Fidget-Spinner) [inline Python/optimizer_analysis.c:279]: Could you delete `is_const` and `get_const` (way below) please? (And replace all their usages in the DSL with these.)

File: Python/tier2_redundancy_eliminator_bytecodes.c [hunk 3]
```diff
@@ -229,10 +230,40 @@ dummy_func(void) {
...
+        OUT_OF_SPACE_IF_NULL(null = sym_new_null(ctx));
+        attr = NULL;
+        if (this_instr[-1].opcode == _NOP) {
+            assert(sym_is_const(owner));
+            PyModuleObject *mod = (PyModuleObject *)sym_get_const(owner);
+            assert(PyModule_CheckExact(mod));
+            PyObject *dict = mod->md_dict;
+            PyObject *res = global_to_const(this_instr, dict);
+            if (res != NULL) {
+                this_instr[-1].opcode = _POP_TOP;
```
Comment (Fidget-Spinner) [inline Python/tier2_redundancy_eliminator_bytecodes.c:262]: We should have a unified way of handling zapping constants in the future, rather than assigning `_NOP` to a guard, `_POP_TOP` to the following, then hoping it gets peepholed away. That can be a future PR though.

**Review by Fidget-Spinner — COMMENTED** [review]
Generally looks good, just one change requested. Thanks!

File: Python/tier2_redundancy_eliminator_bytecodes.c [hunk 4]
```diff
@@ -229,10 +230,40 @@ dummy_func(void) {
...
+        }
+    }
+
     op(_LOAD_ATTR_MODULE, (index/1, owner -- attr, null if (oparg & 1))) {
-        _LOAD_ATTR_NOT_NULL
         (void)index;
-        (void)owner;
+        OUT_OF_SPACE_IF_NULL(null = sym_new_null(ctx));
+        attr = NULL;
+        if (this_instr[-1].opcode == _NOP) {
```
Comment (Fidget-Spinner) [inline Python/tier2_redundancy_eliminator_bytecodes.c:254]: Could you also add a comment here please explaining:
`// No _CHECK_ATTR_MODULE preceding -- means the global was promoted to a constant`

Otherwise this is a little cryptic

File: Python/optimizer_analysis.c [hunk 5]
```diff
@@ -413,18 +415,18 @@ globals_watcher_callback(PyDict_WatchEvent event, PyObject* dict,
...
 global_to_const(_PyUOpInstruction *inst, PyObject *obj)
 {
-    assert(inst->opcode == _LOAD_GLOBAL_MODULE || inst->opcode == _LOAD_GLOBAL_BUILTINS);
+    assert(inst->opcode == _LOAD_GLOBAL_MODULE || inst->opcode == _LOAD_GLOBAL_BUILTINS || inst->opcode == _LOAD_ATTR_MODULE);
     assert(PyDict_CheckExact(obj));
     PyDictObject *dict = (PyDictObject *)obj;
     assert(dict->ma_keys->dk_kind == DICT_KEYS_UNICODE);
     PyDictUnicodeEntry *entries = DK_UNICODE_ENTRIES(dict->ma_keys);
     assert(inst->operand <= UINT16_MAX);
     PyObject *res = entries[inst->operand].me_value;
```
Comment (gvanrossum) [inline Python/optimizer_analysis.c]: Would it be worth to also have a range check on the index clipped by the size of the dict? Or has that already been done? Or would it be too expensive? Even just an assert would reassure me.

File: Python/optimizer_analysis.c [hunk 6]
```diff
@@ -696,6 +700,21 @@ remove_unneeded_uops(_PyUOpInstruction *buffer, int buffer_size)
...
+                }
+                if (last->opcode == _LOAD_CONST_INLINE  ||
+                    last->opcode == _LOAD_CONST_INLINE_BORROW ||
+                    last->opcode == _LOAD_FAST ||
+                    last->opcode == _COPY
+                ) {
+                    last->opcode = _NOP;
+                    buffer[pc].opcode = NOP;
+                }
+            }
```
Comment (gvanrossum) [inline Python/optimizer_analysis.c]: Isn't there a `break` missing here? (Maybe it's time to start putting blank lines between the case blocks, so this kind of thing may stand out a bit more.)

File: Python/optimizer_analysis.c [hunk 7]
```diff
@@ -709,9 +728,6 @@ remove_unneeded_uops(_PyUOpInstruction *buffer, int buffer_size)
                 if (_PyUop_Flags[opcode] & HAS_ERROR_FLAG) {
                     needs_ip = true;
                 }
-                if (opcode == _PUSH_FRAME) {
-                    needs_ip = true;
-                }
```
Comment (gvanrossum) [inline Python/optimizer_analysis.c:714]: Just to confirm, you're removing this because it's got the HAS_ESCAPES_FLAG, right?

File: Python/tier2_redundancy_eliminator_bytecodes.c [hunk 8]
```diff
@@ -229,10 +230,41 @@ dummy_func(void) {
         (void)owner;
     }
 
+    op(_CHECK_ATTR_MODULE, (dict_version/2, owner -- owner)) {
+        (void)dict_version;
+        if (sym_is_const(owner)) {
+            PyModuleObject *mod = (PyModuleObject *)sym_get_const(owner);
+            if (PyModule_CheckExact(mod)) {
```
Comment (gvanrossum) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: I feel a tad uncomfortable with casting to `PyModuleObject *` *before* the `PyModule_CheckExact()` call (even though I know it'll work fine here, it's a lie until that check passes, which it may not).

(Also below in the next op.)

File: Python/tier2_redundancy_eliminator_bytecodes.c [hunk 9]
```diff
@@ -229,10 +230,41 @@ dummy_func(void) {
...
-        (void)owner;
+        OUT_OF_SPACE_IF_NULL(null = sym_new_null(ctx));
+        attr = NULL;
+        if (this_instr[-1].opcode == _NOP) {
+            // Preceding _CHECK_ATTR_MODULE was removed: mod is const and dict is watched.
+            assert(sym_is_const(owner));
+            PyModuleObject *mod = (PyModuleObject *)sym_get_const(owner);
+            assert(PyModule_CheckExact(mod));
+            PyObject *dict = mod->md_dict;
+            PyObject *res = global_to_const(this_instr, dict);
```
Comment (gvanrossum) [inline Python/tier2_redundancy_eliminator_bytecodes.c]: I just realized that the name `global_to_const()` doesn't advertise enough that it actually **modifies the instruction**. Maybe the function name should have a transitive verb in it, e.g. `convert_load_global_to_const()` or `convert_global_to_const()`?

File: Python/tier2_redundancy_eliminator_bytecodes.c [hunk 10]
```diff
@@ -229,10 +230,41 @@ dummy_func(void) {
...
+            if (res != NULL) {
+                this_instr[-1].opcode = _POP_TOP;
```
Comment (gvanrossum) [inline Python/tier2_redundancy_eliminator_bytecodes.c:262]: Hm, I thought we already know that the previous instruction is a `_POP_TOP` -- we check for it on L253 above and only enter this code when it is. Maybe we don't need to set it at all? If there's something you're worried about, maybe use an `assert()` instead?

File: Python/tier2_redundancy_eliminator_bytecodes.c [hunk 11]
```diff
@@ -229,10 +230,41 @@ dummy_func(void) {
...
+                OUT_OF_SPACE_IF_NULL(attr = sym_new_const(ctx, res));
+            }
+        }
+        if (attr == NULL) {
```
Comment (gvanrossum) [inline Python/tier2_redundancy_eliminator_bytecodes.c:266]: I could use a comment explaining that this happens whenever we didn't convert the instruction to a load-constant.

**Review by gvanrossum — COMMENTED** [review]
I have some clarity concerns (based on how long it took me to follow what's happening here) but the basics look good, other than (I think) a missing `break`.

File: Python/optimizer_analysis.c (hunk 5 above)
Comment (markshannon) [reply Python/optimizer_analysis.c]: I'll add a check. It is cheap.
I *think* that it is safe without the check, but I can't prove it.

File: Python/optimizer_analysis.c (hunk 6 above)
Comment (markshannon) [reply Python/optimizer_analysis.c]: Yes, there is a break missing. The compiler usually warns about this. I don't know why it didn't it this case.

File: Python/optimizer_analysis.c (hunk 7 above)
Comment (markshannon) [reply Python/optimizer_analysis.c:714]: Yes

File: Python/tier2_redundancy_eliminator_bytecodes.c (hunk 10 above)
Comment (markshannon) [reply Python/tier2_redundancy_eliminator_bytecodes.c:262]: This previous instruction is a `_NOP` (see line 254), we need to pop the module off the stack, so we convert the previous `_NOP` to `_POP_TOP`.

File: Python/tier2_redundancy_eliminator_bytecodes.c (hunk 11 above)
Comment (markshannon) [reply Python/tier2_redundancy_eliminator_bytecodes.c:266]: Added

File: Python/optimizer_analysis.c [hunk 12]
```diff
@@ -696,6 +703,22 @@ remove_unneeded_uops(_PyUOpInstruction *buffer, int buffer_size)
                 }
                 last_set_ip = pc;
                 break;
+            case _POP_TOP:
```
Comment (Fidget-Spinner) [inline Python/optimizer_analysis.c:706]: We need to unify the zapping of useless instructions along with constants. They are logically the same. We can do that in another PR.

File: Python/tier2_redundancy_eliminator_bytecodes.c [hunk 13]
```diff
@@ -229,10 +230,43 @@ dummy_func(void) {
...
+    op(_CHECK_ATTR_MODULE, (dict_version/2, owner -- owner)) {
+        (void)dict_version;
+        if (sym_is_const(owner)) {
+            PyObject *cnst = sym_get_const(owner);
+            if (PyModule_CheckExact(cnst)) {
+                PyModuleObject *mod = (PyModuleObject *)cnst;
+                PyObject *dict = mod->md_dict;
+                uint64_t watched_mutations = get_mutations(dict);
+                if (watched_mutations < _Py_MAX_ALLOWED_GLOBALS_MODIFICATIONS) {
+                    PyDict_Watch(GLOBALS_WATCHER_ID, dict);
```
Comment (Fidget-Spinner) [inline Python/tier2_redundancy_eliminator_bytecodes.c:242]: This can't fail because `GLOBALS_WATCHER_ID` is a reserved ID right?

**Review by Fidget-Spinner — APPROVED** [review]
LGTM I just have a few questions

File: Python/tier2_redundancy_eliminator_bytecodes.c (hunk 13 above)
Comment (markshannon) [reply Python/tier2_redundancy_eliminator_bytecodes.c:242]: Correct

File: Python/optimizer_analysis.c (hunk 12 above)
Comment (markshannon) [reply Python/optimizer_analysis.c:706]: Probably best done in another pass, but we can worry about that later.

File: Python/tier2_redundancy_eliminator_bytecodes.c (hunk 10 above)
Comment (gvanrossum) [reply Python/tier2_redundancy_eliminator_bytecodes.c:262]: D’oh. I confused the two.
